src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationManager.cs

```diff
@@ -259,7 +234,78 @@ IEnumerator IEnumerable.GetEnumerator()
             }
         }
 
-        private class ConfigurationBuilderProperties : IDictionary<string, object>
+        private sealed class ProviderManager : IDisposable
+        {
+            private readonly object _replaceProvidersLock = new object();
+            private ProvidersReference _refCountedProviders = new(new List<IConfigurationProvider>());
+
+            public IEnumerable<IConfigurationProvider> Providers => _refCountedProviders.Providers;
```

Comment:
You could be accessing disposed providers here.  That's ok?

Reply:
> You could be accessing disposed providers here. That's ok?

It's not completely ok. This was also a problem before this change because we couldn't keep the lock for as long as an app had a reference to the exposed `IEnumerable<IConfigurationProvider>`. I have this comment in the one place we reference this.

```
        // We cannot track the duration of the reference to the providers if this property is used.
        // If a configuration source is removed after this is accessed but before it's completely enumerated,
        // this may allow access to a disposed provider.
        IEnumerable<IConfigurationProvider> IConfigurationRoot.Providers => _providerManager.Providers;
```

I can add a comment here too to explain this is specifically to support `IConfigurationRoot.Providers`.

> Or is the idea that you should only ever be accessing this property when you're holding the reference returned from GetReference? In that case, consider asserting here that its ref count > 0.

That's the idea for the more-commonly-used `ProvidersReference.Providers`. I can add an assert there.